File: src/checkers-python/StudentAI.py
import math
import random
from BoardClasses import Move
from BoardClasses import Board
#The following part should be completed by students.
#Students can modify anything except the class name and exisiting functions and varibles.

class MCTSNode:
    def __init__(self, parent, move, player_to_move):
        self.parent = parent            # parent node
        self.move = move                # move applied at parent to reach this node
        self.player = player_to_move    # side to move at this node
        self.children = []              # list of MCTSNode
        self.untried = []               # moves not expanded yet
        self.visits = 0                 
        self.wins = 0.0                 # wins from ROOT player's POV

class StudentAI():

    def __init__(self,col,row,p):
        self.col = col
        self.row = row
        self.p = p
        self.board = Board(col,row,p)
        self.board.initialize_game()
        self.color = ''
        self.opponent = {1:2,2:1}

    # ---------------------------------------------
    # IS CAPTURE
    # ---------------------------------------------
    def _is_capture_move(self, mv : Move) -> bool:
        """
        Checks if a Move is a capture or not
        
        Args:
            mv (Move): The move to check
        
        Returns:
            bool: Capture move or not
        """

        seq = getattr(mv, "seq", [])
        for (r1, c1), (r2, c2) in zip(seq, seq[1:]):
            if abs(r2 - r1) == 2 and abs(c2 - c1) == 2:
                return True
        return False

    # ---------------------------------------------
    # LEGAL MOVES
    # ---------------------------------------------
    def _legal_moves_flat(self, color) -> list:
        """
        Returns all possible moves of a player shuffled, with a preference towards
        capture moves over quiet moves.

        Args:
            color (int): The color of the player
        
        Returns:
            list: List of all possible moves, flattened
        """

        groups = self.board.get_all_possible_moves(color)
        moves = [m for g in groups for m in g]
        cap = [m for m in moves if self._is_capture_move(m)]
        quiet = [m for m in moves if not self._is_capture_move(m)]
        random.shuffle(cap)
        random.shuffle(quiet)
        return cap + quiet
    
    # ---------------------------------------------
    # WIN CHECK
    # ---------------------------------------------
    def _winner_or_none(self):
        """
        Checks if a player has won or not

        Returns:
            int: 1 if player 1 won, 2 if player 2 won, 
                 None if no one has won
        """

        # return 1 or 2 if someone has won, else None
        if self.board.is_win(1):
            return 1
        if self.board.is_win(2):
            return 2
        return None
    
    # ---------------------------------------------
    # UCT SELECT
    # ---------------------------------------------
    def _uct_select_child(self, node, c=1.4):
        """
        Returns the UCT value of a MCTS Node

        Args:
            node (MCTSNode): node to calculate UCT for
            c (float): exploration factor
        
        Returns:
            UCT value of node
        """

        logN = math.log(node.visits + 1)
        def uct(ch):
            if ch.visits == 0:
                return float("inf")
            exploit = ch.wins / ch.visits
            explore = c * math.sqrt(logN / ch.visits)
            return exploit + explore
        return max(node.children, key = uct)
    
    # ---------------------------------------------
    # EXPANSION
    # ---------------------------------------------
    def _expand(self, node):
        if not node.untried:
            node.untried = self._legal_moves_flat(node.player)
        if not node.untried:
            return node # terminal
        mv = node.untried.pop()
        self.board.make_move(mv, node.player)
        child = MCTSNode(node, mv, self.opponent[node.player])
        node.children.append(child)
        return child
    
    # ---------------------------------------------
    # ROLLOUT
    # ---------------------------------------------
    def _rollout(self, player, max_plies=20):
        local = []

        for _ in range(max_plies):
            # terminal?
            w = self._winner_or_none()
            if w:
                for mv, col in reversed(local): self.board.undo()
                return w

            # moves
            moves = self._legal_moves_flat(player)
            if not moves:
                loser = player
                winner = self.opponent[loser]
                for mv, col in reversed(local): self.board.undo()
                return winner

            # FULL CAPTURE POLICY:
            capture_moves = [m for m in moves if self._is_capture_move(m)]
            if capture_moves:
                # pick the longest capture chain
                max_len = max(len(getattr(mv, "seq", [])) for mv in capture_moves)
                best_caps = [mv for mv in capture_moves if len(getattr(mv, "seq", [])) == max_len]
                mv = random.choice(best_caps)
            else:
                mv = random.choice(moves)

            self.board.make_move(mv, player)
            local.append((mv, player))
            player = self.opponent[player]

        # static fallback (correct perspective)
        score = self.evaluate_color(self.color)
        winner = self.color if score >= 0 else self.opponent[self.color]

        for mv, col in reversed(local): self.board.undo()
        return winner
    
    # ---------------------------------------------
    # BACKPROP
    # ---------------------------------------------
    def _backprop(self, path, winner, root_player):
        for node in path:
            node.visits += 1
            if winner == root_player:
                node.wins += 1.0

    # ---------------------------------------------
    # MCTS SEARCH
    # ---------------------------------------------
    def _mcts_search(self, iterations = 300):
        root = MCTSNode(parent=None, move=None, player_to_move=self.color)
        root.untried = self._legal_moves_flat(root.player)

        if not root.untried:
            return Move([])

        for _ in range(iterations):
            node = root
            path = [node]
            applied = []   # moves applied on self.board in this iteration

            # -------- SELECTION --------
            while not node.untried and node.children:
                node = self._uct_select_child(node)
                if node.move is not None:
                    # apply node's move (chosen by parent.player)
                    self.board.make_move(node.move, node.parent.player)
                    applied.append((node.move, node.parent.player))
                path.append(node)

            # -------- EXPANSION --------
            if node.untried:
                mv = node.untried.pop()
                self.board.make_move(mv, node.player)
                applied.append((mv, node.player))
                child = MCTSNode(node, mv, self.opponent[node.player])
                node.children.append(child)
                node = child
                path.append(node)

            # -------- ROLLOUT --------
            winner = self._rollout(node.player)

            # -------- UNDO SELECTION + EXPANSION --------
            for mv, col in reversed(applied):
                self.board.undo()

            # -------- BACKPROP --------
            self._backprop(path, winner, root_player=self.color)

        # choose move with most visits
        best_child = max(root.children, key=lambda ch: ch.visits)
        return best_child.move
    
    # ---------------------------------------------
    # EVALUATION
    # ---------------------------------------------
    def evaluate_color(self, color):
        """
        Returns a static evaluation of the board from the perspective of `color`.
        Positive score = good for `color`.
        Negative score = bad for `color`.
        """

        opp = self.opponent[color]

        my_score = opp_score = 0

        for i in range(self.board.row):
            for j in range(self.board.col):
                checker = self.board.board[i][j]
                if not checker:
                    continue

                is_king = checker.is_king

                if checker.color == color:
                    my_score += 4 if is_king else 1
                elif checker.color == opp:
                    opp_score += 4 if is_king else 1

        # mobility
        my_moves = self.board.get_all_possible_moves(color)
        opp_moves = self.board.get_all_possible_moves(opp)
        my_score += 0.1 * sum(len(g) for g in my_moves)
        opp_score += 0.1 * sum(len(g) for g in opp_moves)

        return my_score - opp_score

    def get_move(self, move):
        # Determine your assigned color correctly.
        if self.color == '':
            if move:
                # Opponent played first -> you are Player 2 (White)
                self.color = 2
            else:
                # No move history -> you are Player 1 (Black)
                self.color = 1

        # Apply opponent move
        if move:
            self.board.make_move(move, self.opponent[self.color])

        # Run MCTS on the correct root color
        best = self._mcts_search()

        self.board.make_move(best, self.color)
        return best

File: src/checkers-cpp/StudentAI.cpp
#include "StudentAI.h"

//The following part should be completed by students.
//The students can modify anything except the class name and exisiting functions and varibles.
StudentAI::StudentAI(int col,int row,int p)
	:AI(col, row, p)
{
    board = Board(col,row,p);
    board.initializeGame();
    player = 2;
}

Move StudentAI::GetMove(Move move)
{
    if (move.seq.empty())
    {
        player = 1;
    }
    else{
        board.makeMove(move,player == 1?2:1);
    }

    // set up MCTS Node for root node
    MCTSNode *root = new MCTSNode();
    root->parent = nullptr;
    root->currPlayer = player;
    root->untriedMoves = flatten(board, player);
    root->visits = 0;
    root->wins = 0.0;
    root->children.clear();

    for(int i = 0; i < MAX_ITERATIONS; ++i){
        Board simBoard = board;
        MCTSNode *node = root;
        int currentPlayer = root->currPlayer;

        // selection phase
        while(node->untriedMoves.empty() && !node->children.empty()){
            MCTSNode *child = selectChild(node);

            // apply move from child to simBoard
            Move m = child->actionToGetHere;
            simBoard.makeMove(m, currentPlayer);

            // switch player
            currentPlayer = (currentPlayer == 1 ? 2 : 1);

            node = child;
        }

        // expansion phase
        if(!node->untriedMoves.empty()){
            node = expandNode(node, simBoard, currentPlayer);
        }

        // rollout phase
        double result = rollout(simBoard, currentPlayer, player);

        // back propagate result back up
        backprop(node, result);
    }

    MCTSNode *bestChild = nullptr;
    int mostVisits = -1;

    for(auto child : root->children){
        if(child->visits > mostVisits){
            mostVisits = child->visits;
            bestChild = child;
        }
    }

    Move bestMove;

    // fallback if no MCTS expansion available
    if (bestChild == nullptr) {
        // either return an empty Move or pick any legal move from the raw board
        vector<Move> legal = flatten(board, player);
        if (!legal.empty()){
            // choose first move, or make it random?
            bestMove = legal[0];
        }
        else{
            return Move();        // no moves, so game is over
        }
    } 
    else{
        // select the move that made bestChild
        bestMove = bestChild->actionToGetHere;
    }

    // play move
    board.makeMove(bestMove, player);

    // return best move
    return bestMove;

}

// ---------------------------------------------
// FLATTEN MOVES
// ---------------------------------------------
vector<Move> flatten(Board &board, int color){ // 1 = black, 2 = white
    vector<vector<Move>> moves = board.getAllPossibleMoves(color);
    // check if moves is empty
    if(moves.empty()){
        return vector<Move>();
    }

    vector<Move> flattened = moves[0];
    int size = moves.size();
    // go through and insert moves
    for(int i = 1; i < size; i++){
        flattened.insert(flattened.end(), moves[i].begin(), moves[i].end());
    }

    return flattened;
}

// ---------------------------------------------
// CHECK WINNER
// ---------------------------------------------
int isThereWinner(Board &board, int lastPlayer){
    int r = board.isWin(lastPlayer);

    if(r == 1 || r == 2) { return r; }     // winner id
    else if (r == -1) { return 0; }         // tie 
    else { return -1; }                     // game continues
}

// ---------------------------------------------
// UCT (Upper Confidence for Trees)
// ---------------------------------------------
double UCT(const MCTSNode &node, int parentVisits, double C){
    // if node is unvisited, UCT is infinity
    if(node.visits == 0){
        return INFINITY;
    }
    double winRate = double(node.wins) / double(node.visits);
    double exploration = C * sqrt((log(parentVisits + 1) / node.visits));
    
    return winRate + exploration;
}

// ---------------------------------------------
// EVALUATION HEURISTIC
// ---------------------------------------------
double evaluatePlayer(Board &board, int player){
    // establish colors of player and opponent
    string playerColor = (player == 1 ? "B" : "W");
    string opponentColor = (player == 1 ? "W" : "B");

    // establish opponent player
    int opponent = (player == 1 ? 2 : 1);

    // establish player and opponent's scores
    double playerScore = 0.0;
    double opponentScore = 0.0;

    int rows = board.row;
    int cols = board.col;

    // material score
    for(int i = 0; i < rows; ++i){
        for(int j = 0; j < cols; ++j){
            Checker checker = board.board[i][j];
            // no checker there, continue
            if(checker.color == ".") { continue; }

            // check if checker is a king
            bool isKing = checker.isKing;

            if(checker.color == playerColor){
                playerScore += (isKing ? 0.4 : 0.1);
            }
            else if(checker.color == opponentColor){
                opponentScore += (isKing ? 0.4 : 0.1);
            }
        }
    }

    // mobility score
    vector<Move> playerMoves = flatten(board, player);
    vector<Move> opponentMoves = flatten(board, opponent);
    playerScore += 0.1 * playerMoves.size();
    opponentScore += 0.1 * opponentMoves.size();

    double diff = playerScore - opponentScore;
    double big = 0.5;
    double small = 0.2;
    
    if (diff >= big)            { return 0.9; } // big lead
    else if (diff >= small)     { return 0.7; } // small lead
    else if (diff > -small)     { return 0.5; } // roughly equal
    else if (diff > -big)       { return 0.3; } // small disadvantage
    else                        { return 0.1; } // big disadvantage
}

// ---------------------------------------------
// SELECT CHILD
// ---------------------------------------------
MCTSNode* selectChild(MCTSNode *node){
    double maxUCT = -INFINITY;
    MCTSNode *bestChild = nullptr;

    int parentVisits = node->visits;

    for(MCTSNode *child : node->children){
        double uct = UCT(*child, parentVisits);
        if(uct > maxUCT){
            maxUCT = uct;
            bestChild = child;
        }
    }
    return bestChild;
}

// ---------------------------------------------
// EXPAND NODE
// ---------------------------------------------
MCTSNode* expandNode(MCTSNode *node, Board &simBoard, int &currentPlayer){
    Move move = node->untriedMoves.back(); // last move
    node->untriedMoves.pop_back(); // remove move

    // apply move to simBoard
    simBoard.makeMove(move, currentPlayer);

    MCTSNode *child = new MCTSNode();
    child->parent = node;
    child->actionToGetHere = move;

    // switch players
    currentPlayer = (currentPlayer == 1 ? 2 : 1);
    child->currPlayer = currentPlayer;

    // moves from new position
    child->untriedMoves = flatten(simBoard, child->currPlayer);
    child->visits = 0;
    child->wins = 0.0;

    node->children.push_back(child);
    return child;
}

// ---------------------------------------------
// ROLLOUT
// ---------------------------------------------
double rollout(Board &simBoard, int currentPlayer, int myPlayer){
    for(int i = 0; i < MAX_ROLLOUT_PLIES; ++i){
        // generate all legal moves for currentPlayer
        vector<Move> legalMoves = flatten(simBoard, currentPlayer);
        int legalMovesSize = legalMoves.size();

        // if no legal moves for myPlayer...
        if(legalMovesSize == 0 && currentPlayer == myPlayer){
            // ... myPlayer loses, return 0.0
            return 0.0;
        }
        // if no legal moves for myPlayer's opponent...
        else if(legalMovesSize == 0 && currentPlayer != myPlayer){
            // ... myPlayer wins, return 1.0
            return 1.0;
        }

        // choose random move and play it
        int randIndex = std::rand() % legalMoves.size();
        Move randomMove = legalMoves[randIndex];
        simBoard.makeMove(randomMove, currentPlayer);

        // check if node is at a winning state
        int winner = isThereWinner(simBoard, currentPlayer);
        if(winner != -1){
            if(winner == myPlayer) { return 1.0; } // if myPlayer won, return 1.0
            else if(winner == 0) { return 0.5; } // tie case
            else { return 0.0; }
        }

        // switch to opponent
        currentPlayer = (currentPlayer == 1 ? 2 : 1);
    }

    // max-depth reached, so evaluate board based on myPlayer
    return evaluatePlayer(simBoard, myPlayer);
}

// ---------------------------------------------
// BACK PROPAGATION
// ---------------------------------------------
void backprop(MCTSNode *node, double result){
    while(node != nullptr){
        node->visits += 1;
        node->wins += result;
        node = node->parent;
    }
}

